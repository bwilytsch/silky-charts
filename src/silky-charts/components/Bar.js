import React, { useState, useRef, useEffect, useMemo } from 'react';
import identity from 'ramda/src/identity';
import { select as d3Select } from 'd3-selection';
import { max as d3Max } from 'd3-array';
import { axisBottom as d3AxisBottom, axisLeft as d3AxisLeft } from 'd3-axis';
import { Axis, BarDatum, Grid, SVG } from './styled';
import {
  allDate,
  drawGrid,
  extendXPath,
  getBaseColor,
  getDivergence,
  getId,
  getSize,
  getXScale,
  getYScale,
  rotateXLabels,
  setupData,
  valueFor,
} from '../utils';
import {
  SCALE_TIME,
  SCALE_BAND,
  SCALE_LINEAR,
  THEME,
  TICKS,
  MARGIN,
  ASPECT_RATIO,
  SIZE,
} from '../utils/constants';

const Bar = ({
  aspectRatio = ASPECT_RATIO,
  data: chartData,
  showValue,
  showDivergence,
  grid,
  horizontal,
  margin = MARGIN,
  onClick = identity,
  onMouseEnter = identity,
  onMouseLeave = identity,
  width: svgWidth = undefined,
  height: svgHeight = undefined,
  theme = THEME,
  ticks = TICKS,
  xAxisLabelRotation,
  xAxisLabelRotationValue = -50,
}) => {
  const svgRef = useRef();
  const [{ width, height, isSizeSet }, setSize] = useState(SIZE);
  const [currentValue, setCurrentValue] = useState(null);
  const [isDates, data] = useMemo(() => setupData(chartData), chartData);
  const [id] = useState(getId('bar'));

  const xScale = getXScale(
    isDates ? SCALE_TIME : SCALE_BAND,
    data,
    width,
    true
  );
  const yScale = getYScale(
    SCALE_LINEAR,
    d3Max(data, ({ value }) => value),
    height
  );

  const handleOnMouseEnter = event => {
    setCurrentValue(event.target.getAttribute('value'));
    onMouseEnter(event);
  };

  const handleOnMouseLeave = event => {
    setCurrentValue(null);
    onMouseLeave(event);
  };

  const handleResize = () => {
    const offsetWidth = svgRef.current.parentElement.offsetWidth;
    if ((svgWidth || svgHeight) && !isSizeSet) {
      setSize({
        ...getSize(svgWidth, svgHeight, margin, aspectRatio),
        isSizeSet: true,
      });
    } else if (offsetWidth !== svgWidth - (margin.left + margin.right)) {
      setSize({
        ...getSize(offsetWidth, undefined, margin, aspectRatio),
        isSizeSet: true,
      });
    }
  };

  useEffect(() => {
    handleResize();
    window.addEventListener('resize', handleResize);

    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, []);

  return (
    <SVG
      ref={svgRef}
      identifier={id}
      size={{
        width: svgWidth || width + margin.left + margin.right,
        height: svgHeight || height + margin.top + margin.bottom,
      }}
    >
      <g
        className="container"
        transform={`translate(${margin.left}, ${margin.top})`}
      >
        {grid && (
          <Grid
            ref={node =>
              d3Select(node).call(
                drawGrid(horizontal, xScale, height, yScale, width, ticks)
              )
            }
          />
        )}

        <Axis
          axis="x"
          position={{ x: 0, y: height }}
          ref={node => {
            d3Select(node).call(d3AxisBottom(xScale));
            isDates && extendXPath(id, width);
            xAxisLabelRotation && rotateXLabels(id, xAxisLabelRotationValue);
          }}
        />
        <Axis
          axis="y"
          ref={node => d3Select(node).call(d3AxisLeft(yScale).ticks(ticks))}
        />

        <g className="data">
          {data.map(({ name, value }, idx) => (
            <BarDatum
              key={idx}
              showValue={showValue}
              datum={{
                name,
                value:
                  showValue && showDivergence && currentValue
                    ? getDivergence(value, currentValue)
                    : value,
              }}
              x={
                isDates
                  ? xScale(name) - valueFor('x', width, data.length)
                  : xScale(name)
              }
              y={yScale(value)}
              width={
                isDates
                  ? valueFor('width', width, data.length)
                  : xScale.bandwidth()
              }
              height={height - yScale(value)}
              color={getBaseColor(theme)}
              onClick={onClick}
              onMouseEnter={handleOnMouseEnter}
              onMouseLeave={handleOnMouseLeave}
            />
          ))}
        </g>
      </g>
    </SVG>
  );
};

export default Bar;
