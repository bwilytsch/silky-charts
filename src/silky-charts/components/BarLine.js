import React, { useState, useRef, useEffect, useMemo } from 'react';
import head from 'ramda/src/head';
import identity from 'ramda/src/identity';
import { select as d3Select } from 'd3-selection';
import { line as d3Line } from 'd3-shape';
import { max as d3Max } from 'd3-array';
import { axisBottom as d3AxisBottom, axisLeft as d3AxisLeft } from 'd3-axis';
import { Axis, StackedBarDatum, Grid, Label, LineDatum, SVG } from './styled';
import {
  buildStack,
  drawGrid,
  extendXPath,
  getId,
  getLineDataForKeys,
  getMaximumValues,
  getSize,
  getXScale,
  getYScale,
  palette,
  rotateXLabels,
  setupData,
  setLineType,
} from '../utils';
import {
  SCALE_TIME,
  SCALE_BAND,
  SCALE_LINEAR,
  SECONDARY_THEME,
  THEME,
  TICKS,
  LINE_TYPE,
  MARGIN,
  ASPECT_RATIO,
  SIZE,
} from '../utils/constants';

const BarLine = ({
  aspectRatio = ASPECT_RATIO,
  data: chartData,
  grid,
  horizontal,
  margin = MARGIN,
  onClick = identity,
  onMouseEnter = identity,
  onMouseLeave = identity,
  stackedKeys = [],
  lineKeys = [],
  lineType = LINE_TYPE,
  lineTypeOption = null,
  width: svgWidth = undefined,
  height: svgHeight = undefined,
  theme = THEME,
  ticks = TICKS,
  secondaryTheme = SECONDARY_THEME,
  xAxisLabelRotation,
  xAxisLabelRotationValue = -50,
  xAxisLabel,
  yAxisLabel,
}) => {
  const svgRef = useRef();
  const [{ width, height, isSizeSet }, setSize] = useState(SIZE);
  const [isDates, data] = useMemo(() => setupData(chartData), chartData);
  const [id] = useState(getId('bar-line'));

  const xScale = getXScale(
    isDates ? SCALE_TIME : SCALE_BAND,
    data,
    width,
    true
  );
  const yScale = getYScale(
    SCALE_LINEAR,
    d3Max(getMaximumValues(stackedKeys, data)),
    height
  );

  const stack = buildStack(stackedKeys);
  const line = d3Line()
    .x(({ name }) =>
      isDates ? xScale(name) : xScale(name) + xScale.bandwidth() / 2
    )
    .y(({ value }) => yScale(value))
    .curve(setLineType(lineType, lineTypeOption));

  const lineData = getLineDataForKeys(lineKeys, data);

  const handleResize = () => {
    const offsetWidth = svgRef.current.parentElement.offsetWidth;
    if ((svgWidth || svgHeight) && !isSizeSet) {
      setSize({
        ...getSize(svgWidth, svgHeight, margin, aspectRatio),
        isSizeSet: true,
      });
    } else if (offsetWidth !== svgWidth - (margin.left + margin.right)) {
      setSize({
        ...getSize(offsetWidth, undefined, margin, aspectRatio),
        isSizeSet: true,
      });
    }
  };

  useEffect(() => {
    handleResize();
    window.addEventListener('resize', handleResize);

    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, []);

  return (
    <SVG
      ref={svgRef}
      identifier={id}
      size={{
        width: svgWidth || width + margin.left + margin.right,
        height: svgHeight || height + margin.top + margin.bottom,
      }}
    >
      <g
        className="silky-charts-container"
        transform={`translate(${margin.left}, ${margin.top})`}
      >
        {grid && (
          <Grid
            ref={node =>
              d3Select(node).call(
                drawGrid(horizontal, xScale, height, yScale, width, ticks)
              )
            }
          />
        )}

        {xAxisLabel && (
          <Label axis="x" margin={margin} width={width} height={height}>
            {xAxisLabel}
          </Label>
        )}

        {yAxisLabel && (
          <Label axis="y" margin={margin} width={width} height={height}>
            {yAxisLabel}
          </Label>
        )}

        <Axis
          axis="x"
          position={{ x: 0, y: height }}
          ref={node => {
            d3Select(node).call(d3AxisBottom(xScale));
            isDates && extendXPath(id, width);
            xAxisLabelRotation && rotateXLabels(id, xAxisLabelRotationValue);
          }}
        />
        <Axis
          axis="y"
          ref={node => d3Select(node).call(d3AxisLeft(yScale).ticks(ticks))}
        />

        <StackedBarDatum
          data={data}
          series={stack(data)}
          isDates={isDates}
          onClick={onClick}
          theme={theme}
          x={xScale}
          y={yScale}
          width={width}
          height={height}
        />

        {lineData.map((datum, idx) => (
          <g className={`${head(datum)['key']}-layer`} key={idx}>
            <LineDatum
              data={datum}
              d={line(datum)}
              isDates={isDates}
              xScale={xScale}
              yScale={yScale}
              color={palette.themes[secondaryTheme].base[idx]}
            />
          </g>
        ))}
      </g>
    </SVG>
  );
};

export default BarLine;
