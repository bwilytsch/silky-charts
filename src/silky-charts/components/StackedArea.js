import React, { useEffect, useMemo, useRef, useState } from 'react';
import identity from 'ramda/src/identity';
import { axisBottom as d3AxisBottom, axisLeft as d3AxisLeft } from 'd3-axis';
import { select as d3Select } from 'd3-selection';
import { area as d3Area } from 'd3-shape';
import { Axis, Grid, Label, Path, SVG } from './styled';
import {
  appendStackedValues,
  buildStack,
  bySeries,
  classify,
  debounce,
  drawGrid,
  getId,
  getMax,
  getStackedMax,
  getSeries,
  getSize,
  getXScale,
  getYScale,
  palette,
  rotateXLabels,
  setLineType,
  setupData,
  toStackedForm,
} from '../utils';
import {
  ASPECT_RATIO,
  LINE_TYPE,
  MARGIN,
  ROTATION,
  SCALE_BAND,
  SCALE_LINEAR,
  SCALE_TIME,
  SIZE,
  THEME,
  TICKS,
} from '../utils/constants';

const StackedArea = ({
  aspectRatio = ASPECT_RATIO,
  data: chartData,
  grid,
  height: svgHeight = undefined,
  isHorizontal,
  lineSeries = [],
  lineType = LINE_TYPE,
  lineTypeOption = null,
  margin = MARGIN,
  onClick = identity,
  onMouseEnter = identity,
  onMouseLeave = identity,
  responsive = false,
  theme = THEME,
  ticks = TICKS,
  width: svgWidth = undefined,
  xAxisLabel,
  xAxisLabelRotation,
  xAxisLabelRotationValue = ROTATION,
  yAxisLabel,
}) => {
  const svgRef = useRef();
  const [id] = useState(getId('stacked-area'));
  const [{ width, height, isSizeSet }, setSize] = useState(SIZE);
  let [isDates, data] = useMemo(() => setupData(chartData), chartData);
  data = useMemo(
    () =>
      appendStackedValues(
        buildStack(getSeries(data))(toStackedForm(data)),
        data
      ),
    data
  );

  const xScale = getXScale(isDates ? SCALE_TIME : SCALE_BAND, data, width);
  const yScale = getYScale(SCALE_LINEAR, getMax(getStackedMax(data)), height);

  const area = d3Area()
    .curve(setLineType(lineType, lineTypeOption))
    .x(({ name }) => xScale(name))
    .y0(({ stackedValues }) => yScale(stackedValues[0]))
    .y1(({ stackedValues }) => yScale(stackedValues[1]));

  const handleSize = () => {
    const offsetWidth = svgRef.current.parentElement.offsetWidth;
    if ((svgWidth || svgHeight) && !isSizeSet) {
      setSize({
        ...getSize(svgWidth, svgHeight, margin, aspectRatio),
        isSizeSet: true,
      });
    } else if (offsetWidth !== svgWidth - (margin.left + margin.right)) {
      setSize({
        ...getSize(offsetWidth, undefined, margin, aspectRatio),
        isSizeSet: true,
      });
    }
  };

  const handleResize = debounce(handleSize)();

  useEffect(() => {
    handleSize();
    responsive && window.addEventListener('resize', handleResize);

    return () => {
      responsive && window.removeEventListener('resize', handleResize);
    };
  }, []);

  return (
    <SVG
      identifier={id}
      size={{
        width: svgWidth || width + margin.left + margin.right,
        height: svgHeight || height + margin.top + margin.bottom,
      }}
      ref={svgRef}
    >
      <g
        className="silky-charts-container"
        transform={`translate(${margin.left}, ${margin.top})`}
      >
        {grid && (
          <Grid
            ref={node =>
              d3Select(node).call(
                drawGrid(isHorizontal, xScale, height, yScale, width, ticks)
              )
            }
          />
        )}

        {xAxisLabel && (
          <Label axis="x" margin={margin} width={width} height={height}>
            {xAxisLabel}
          </Label>
        )}

        {yAxisLabel && (
          <Label axis="y" margin={margin} width={width} height={height}>
            {yAxisLabel}
          </Label>
        )}

        {bySeries(data).map(([series, datum], idx) => (
          <g className={`${classify(series)}-layer`} key={idx}>
            <Path
              chart="stacked-area"
              fillColor={palette.themes[theme].base[idx]}
              d={area(datum)}
              strokeWidth={0}
              onClick={onClick}
              onMouseEnter={onMouseEnter}
              onMouseLeave={onMouseLeave}
            />
          </g>
        ))}

        <Axis
          axis="x"
          position={{ x: 0, y: height }}
          ref={node => {
            d3Select(node).call(d3AxisBottom(xScale));
            xAxisLabelRotation && rotateXLabels(id, xAxisLabelRotationValue);
          }}
        />
        <Axis
          axis="y"
          ref={node => d3Select(node).call(d3AxisLeft(yScale).ticks(ticks))}
        />
      </g>
    </SVG>
  );
};

export default StackedArea;
