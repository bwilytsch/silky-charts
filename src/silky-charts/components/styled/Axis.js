import styled from 'styled-components';

const Axis = styled.g.attrs(({ axis, translate }) => ({
  className: `axis-${axis}`,
  transform: translate && `translate(${translate.x}, ${translate.y})`,
}))``;

export default Axis;
