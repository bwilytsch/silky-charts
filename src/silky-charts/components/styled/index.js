export { default as Axis } from './Axis';
export { default as BarDatum } from './BarDatum';
export { default as Circle } from './Circle';
export { default as Grid } from './Grid';
export { default as LineDatum } from './LineDatum';
export { default as Path } from './Path';
export { default as StackedBarDatum } from './StackedBarDatum';
export { default as SVG } from './SVG';
