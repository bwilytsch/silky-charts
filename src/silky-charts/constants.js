// Color pallete
export const pallete = {
  white: 'rgb(255, 255, 255)', // #FFFFFF
  black: 'rgb(33, 33, 33)', // #212121
  grid: 'rgb(220, 220, 220)', // #DCDCDC
  reference: 'rgb(170, 93, 0)', //#AA5D00
  bg: {
    blue: 'rgb(10, 48, 85)', // #0A3055
    gray: 'rgb(46, 52, 59)', // #2E343B
    green: 'rgb(17, 76, 42)', // #114C2A
    orange: 'rgb(102, 56, 15)', // #66380F
    purple: 'rgb(89, 29, 119)', // #591D77
    red: 'rgb(93, 18, 18)', // #5D1212
    yellow: 'rgb(90, 68, 13)', // #5A440D
  },
  fg: {
    blue: 'rgb(15, 72, 128)', // #0F4880
    gray: 'rgb(79, 90, 101)', // #4F5A65
    green: 'rgb(32, 96, 60)', // #20603C
    orange: 'rgb(116, 78, 46)', // #744E2E
    purple: 'rgb(118, 0, 168)', // #7600A8
    red: 'rgb(139, 0, 0)', // #8B0000
    yellow: 'rgb(85, 83, 68)', // #555344
  },
}
