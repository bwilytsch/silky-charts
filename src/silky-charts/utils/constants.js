// Scales
export const SCALE_BAND = 'band';
export const SCALE_LINEAR = 'linear';
export const SCALE_TIME = 'time';

export const TICKS = 5;

// Themes
export const THEME = 'tealBlue';
export const SECONDARY_THEME = 'pink';
