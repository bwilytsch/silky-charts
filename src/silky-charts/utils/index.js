export { default as allDate } from './allDate';
export { default as drawGrid } from './drawGrid';
export { default as extendXPath } from './extendXPath';
export { default as getDivergence } from './getDivergence';
export { default as getXScale } from './getXScale';
export { default as getYScale } from './getYScale';
export { default as rotateXLabels } from './rotateXLabels';
export { default as themes } from './themes';
