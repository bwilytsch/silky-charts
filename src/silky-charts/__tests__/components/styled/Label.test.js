import React from 'react';
import { Label } from '../../../components/styled';
import { create } from 'react-test-renderer';
import 'jest-styled-components';
import { MARGIN } from '../../../utils/constants';

test('render correctly for x axis', () => {
  const tree = create(
    <Label axis="x" margin={MARGIN} width={10} height={10} />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test('render correctly for y axis', () => {
  const tree = create(
    <Label axis="y" margin={MARGIN} width={10} height={10} />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
