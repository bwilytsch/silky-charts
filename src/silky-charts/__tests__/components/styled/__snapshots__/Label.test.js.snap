// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly for x axis 1`] = `
.c0 {
  text-anchor: middle;
}

<text
  className="x-axis-label c0"
  height={10}
  width={10}
  x={5}
  y={43.333333333333336}
/>
`;

exports[`render correctly for y axis 1`] = `
.c0 {
  text-anchor: middle;
  -webkit-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
}

<text
  className="y-axis-label c0"
  height={10}
  width={10}
  x={-5}
  y={-33.333333333333336}
/>
`;
